### ISSUE TITLE
Bug: [explicit-function-return-type] false positive when returning an object with a typed function property

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8918]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&showAST=es&fileType=.tsx&code=C4TwDgpgBA8gRgKwgY2AdQJbABYBVwQAmAYgK4B2qGA9uVALxQDeAUFFAGYBcUAFAJQMAfFABu1DIQDcLAL4sWHClVpRqxAIapqAJxACe8JKkw58kEsuA06rdjojBSO223aceA4cyjz282SA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroDNYTMvkoB7JrWiJ8saJKKl0URNGhjokcGAC%2BIHUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type ObjectWithTypedFunction = {
  f: () => void;
}

function oFactory(): ObjectWithTypedFunction {
  return {
    f: () => { }
  }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/explicit-function-return-type": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect that the arrow function literal in the factory function should not flag

### Actual Result

it flags.

### Additional Info

There seems to be something specific about function properties on an object. In the following code ([link to playground](https://typescript-eslint.io/play/#ts=5.2.2&showAST=es&fileType=.tsx&code=C4TwDgpgBAKuEBMBiBXAdgY2ASwPZqgF4oAKASiID4oA3XbBAbgChmN8BnYKAMwC5Y8ZOix4CxclSgBvKAF8WzHiJz5eSAIZZcAJxDkBcSMMyqC05lCg6IwFDoKTC1WQuZzWoSFADyAIwArCCwAdWxgAAsjRFRTMSIZS15DIVjRfHdWdjQuKFwBfyDQ8KjUlXjiCyt%2BUgpnGXlM5gRggBsNG15ytQgAD0gsDkLg4DDI6JN0tBIOgHMABgLAkbHS4zSzMgE6BkU%2BgeAh5eLxsrj8Eirk2qlXdzJWZXOCXE1tPQNfY9GSiY34q42OwORJWaoCJwuRpWDxyIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroDNYTMvkoB7JrWiJ8saJKKl0URNGhjokcGAC%2BIHUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)), only the one line flags, whereas the rest correctly do not:
```ts
type TypedFunction = () => void;

const f: TypedFunction = () => { };

function fFactory(): TypedFunction {
  return () => { };
}

type ObjectWithTypedFunction = {
  f: TypedFunction
}

const o: ObjectWithTypedFunction = {
  f: () => { }
}

declare function expectsObjectWithTypedFunction(arg0: ObjectWithTypedFunction): void;

expectsObjectWithTypedFunction({
  f: () => { }
})

function oFactory(): ObjectWithTypedFunction {
  return {
    f: () => { }
//        ^^  <-- Missing return type on function
  }
}
```

Noticed in https://github.com/typescript-eslint/typescript-eslint/pull/8821/files#r1564644417

**Comment by bradzacher** [comment]
IIRC this getting to be a "working as intended" scenario. You can probably dig up some old issues asking about this behaviour.
The reason being that part of the motivation behind this rule is to make it easier to review code without type analysis turned on.

But when you've got a function that has a type, and that function returns an object and that object has properties and those properties might have a strict function return type..... well there's so much indirection there that we consider the rule violated.
Also imagine code like this:
```ts
function foo(): ReturnType {
  // ...
  // 300 lines
  // ...

  return {
    prop: () => {}
  };
}
```

It's just so far removed from the explicitness that it's no longer really clear what it is any more.

**Comment by kirkwaiblinger** [comment]
@bradzacher 
Interesting. Yeah, it looks like maybe it's this code that controls it?
https://github.com/typescript-eslint/typescript-eslint/blob/612875bced1cd5a197a97c58dc9a89bf9e59cd8a/packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts#L83-L94

Whereas my mental model was assuming something like
```diff
 function isTypedParent( 
   parent: TSESTree.Node, 
   callee?: FunctionExpression, 
 ): boolean { 
   return ( 
     isTypeAssertion(parent) || 
     isVariableDeclaratorWithTypeAnnotation(parent) || 
     isPropertyDefinitionWithTypeAnnotation(parent) || 
     isFunctionArgument(parent, callee) || 
+    // checks if return statement is constrained by an explicit return type
+    isTypedReturnStatement(parent) ||
     isTypedJSX(parent) 
   ); 
 } 
```

**Comment by JoshuaKGoldberg** [comment]
👋 Just checking in @kirkwaiblinger, what do you think should happen next?

**Comment by kirkwaiblinger** [comment]
@JoshuaKGoldberg I think let's finish the conversation in #9171 instead and then close or action this issue with the same resolution

**Comment by bradzacher** [comment]
As with https://github.com/typescript-eslint/typescript-eslint/issues/9171#issuecomment-2308655956 - I think we should reject this.

**Comment by kirkwaiblinger** [comment]
Closed per discussion in https://github.com/typescript-eslint/typescript-eslint/issues/9171
